# Python and Pandas Exercise with Dataframes

## Importing Libraries and Reading Data

Make sure that your dataset file and Jupyter notebook are in the same location

In [1]:
import pandas as pd
import numpy as np

In [31]:
# import dataframe from a file
df = pd.read_csv('school_grades.csv')

In [3]:
df

gender class group parental level of education         lunch  \
0      male     group A                 high school      standard   
1    female     group D            some high school  free/reduced   
2      male     group E                some college  free/reduced   
3      male     group B                 high school      standard   
4      male     group E          associate's degree      standard   
..      ...         ...                         ...           ...   
995    male     group C                 high school      standard   
996    male     group D          associate's degree  free/reduced   
997  female     group C            some high school  free/reduced   
998  female     group C                some college      standard   
999    male     group A                some college      standard   

    test preparation course  math score  reading score  writing score  
0                 completed          67             67             63  
1                      none          40             59             55  
2                      none          59             60             50  
3                      none          77             78             68  
4                 completed          78             73             68  
..                      ...         ...            ...            ...  
995                    none          73             70             65  
996               completed          85             91             92  
997                    none          32             35             41  
998                    none          73             74             82  
999               completed          65             60             62  

[1000 rows x 8 columns]

In [4]:
# get the top 10 rows of the dataframe
df.head(10)

gender class group parental level of education         lunch  \
0    male     group A                 high school      standard   
1  female     group D            some high school  free/reduced   
2    male     group E                some college  free/reduced   
3    male     group B                 high school      standard   
4    male     group E          associate's degree      standard   
5  female     group D                 high school      standard   
6  female     group A           bachelor's degree      standard   
7    male     group E                some college      standard   
8    male     group D                 high school      standard   
9    male     group C                some college  free/reduced   

  test preparation course  math score  reading score  writing score  
0               completed          67             67             63  
1                    none          40             59             55  
2                    none          59             60             50  
3                    none          77             78             68  
4               completed          78             73             68  
5                    none          63             77             76  
6                    none          62             59             63  
7               completed          93             88             84  
8                    none          63             56             65  
9                    none          47             42             45

In [5]:
# get the last 7 rows of the df
df.tail(7)

gender class group parental level of education         lunch  \
993  female     group A          associate's degree      standard   
994    male     group E                 high school  free/reduced   
995    male     group C                 high school      standard   
996    male     group D          associate's degree  free/reduced   
997  female     group C            some high school  free/reduced   
998  female     group C                some college      standard   
999    male     group A                some college      standard   

    test preparation course  math score  reading score  writing score  
993                    none          58             54             58  
994               completed          86             82             75  
995                    none          73             70             65  
996               completed          85             91             92  
997                    none          32             35             41  
998                    none          73             74             82  
999               completed          65             60             62

In [6]:
df.sample(5)

gender class group parental level of education         lunch  \
856    male     group C            some high school  free/reduced   
990    male     group D                some college      standard   
139  female     group C            some high school  free/reduced   
220    male     group D                some college      standard   
181  female     group C                some college      standard   

    test preparation course  math score  reading score  writing score  
856                    none          61             57             55  
990                    none          67             55             53  
139                    none          60             72             64  
220                    none          89             90             83  
181                    none          48             52             49

In [7]:
df.shape #(row and col counts)

(1000, 8)

In [8]:
print(f'We have {df.shape[0]} rows and {df.shape[1]} columns')

We have 1000 rows and 8 columns


In [9]:
# check for nulls
df.isna().sum()

gender                         0
class group                    0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

**Observations**
- We have 1000 rows and 8 columns
- There are no nulls in this dataset

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   class group                  1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


In [11]:
# to get a real memory usage value, use deep
df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   class group                  1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 342.5 KB


- real mem usage is 342.5 KB
- we have different data types:
    - 3 integer columns (int64)
    - 5 string columns (object)

In [13]:
# view the data using a specific range - loc
df.loc[200:210]

gender class group parental level of education         lunch  \
200  female     group D          associate's degree      standard   
201    male     group E          associate's degree      standard   
202    male     group C                some college      standard   
203  female     group B            some high school      standard   
204  female     group D          associate's degree  free/reduced   
205    male     group B            some high school      standard   
206  female     group A             master's degree      standard   
207    male     group C                some college      standard   
208  female     group A                some college      standard   
209  female     group B          associate's degree      standard   
210  female     group D                 high school  free/reduced   

    test preparation course  math score  reading score  writing score  
200               completed          50             63             65  
201                    none          55             54             47  
202                    none          77             64             70  
203               completed          49             64             67  
204                    none          66             83             82  
205               completed          77             74             75  
206               completed          74             78             84  
207                    none          85             78             81  
208                    none          77             86             84  
209                    none          64             62             66  
210               completed          61             69             68

In [15]:
# get specific col list - with a sample of 10
df[['gender', 'lunch', 'writing score']].sample(10)

gender         lunch  writing score
560  female      standard             57
772  female      standard             78
821    male      standard             64
483  female  free/reduced             82
901  female  free/reduced             68
204  female  free/reduced             82
356  female  free/reduced             68
767  female      standard             83
878  female      standard             63
286  female      standard             84

In [16]:
# row range and col range
df.iloc[100:111, 3:6]

lunch test preparation course  math score
100  free/reduced               completed          54
101      standard               completed          75
102      standard                    none          79
103      standard                    none          66
104      standard                    none          47
105      standard                    none          70
106      standard                    none          82
107      standard                    none          56
108      standard                    none          92
109  free/reduced                    none          30
110      standard               completed          79

In [20]:
df.iloc[100:111][['gender', 'reading score']]

gender  reading score
100    male             61
101    male             78
102    male             78
103    male             67
104  female             58
105  female             78
106    male             74
107    male             38
108    male             78
109    male             33
110  female             88

In [21]:
# drop columns and create a new dataset
df2 = df.drop(columns=['gender','lunch'], axis=1) # for compatibility of any pandas version use axis=1 
df2.head()

class group parental level of education test preparation course  math score  \
0     group A                 high school               completed          67   
1     group D            some high school                    none          40   
2     group E                some college                    none          59   
3     group B                 high school                    none          77   
4     group E          associate's degree               completed          78   

   reading score  writing score  
0             67             63  
1             59             55  
2             60             50  
3             78             68  
4             73             68

In [17]:
# df.info alternative for data types
df.dtypes

gender                         object
class group                    object
parental level of education    object
lunch                          object
test preparation course        object
math score                      int64
reading score                   int64
writing score                   int64
dtype: object

In [22]:
# changing data types
df['reading score'] = df['reading score'].astype('int16')

In [23]:
df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   class group                  1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int16 
 7   writing score                1000 non-null   int64 
dtypes: int16(1), int64(2), object(5)
memory usage: 336.6 KB


In [ ]:
df[['math score', 'reading score', 'writing score']] = df[['math score', 'reading score', 'writing score']].astype('int16')

In [33]:
# method 2 of converting multiple columns based on data type
num_col_list = df.select_dtypes(include='int').columns
num_col_list

Index(['math score', 'reading score', 'writing score'], dtype='object')

In [ ]:
df[num_col_list] = df[num_col_list].astype('int16')

In [ ]:
num_col_list = df.select_dtypes(exclude='object').columns

## Statistical Functions

In [24]:
# get the sum of numerical columns
df.sum(numeric_only=True)

math score       66396
reading score    69002
writing score    67738
dtype: int64

In [25]:
df.max(numeric_only=True)

math score       100
reading score    100
writing score    100
dtype: int64

In [26]:
df.describe() # statistical measures of the numerical columns in the dataset

math score  reading score  writing score
count  1000.000000    1000.000000    1000.000000
mean     66.396000      69.002000      67.738000
std      15.402871      14.737272      15.600985
min      13.000000      27.000000      23.000000
25%      56.000000      60.000000      58.000000
50%      66.500000      70.000000      68.000000
75%      77.000000      79.000000      79.000000
max     100.000000     100.000000     100.000000

In [30]:
# for a specific column
df['reading score'].mean()

69.002

In [28]:
# export a slice of data into json
df.iloc[100:111, 3:6].to_json('my_data.json')

## Renaming Columns

### method 1

In [34]:
#method 1
df2 = df.rename(columns={'reading score': 'reading grade', 'math score': 'math grade'})
df2.head()

gender class group parental level of education         lunch  \
0    male     group A                 high school      standard   
1  female     group D            some high school  free/reduced   
2    male     group E                some college  free/reduced   
3    male     group B                 high school      standard   
4    male     group E          associate's degree      standard   

  test preparation course  math grade  reading grade  writing score  
0               completed          67             67             63  
1                    none          40             59             55  
2                    none          59             60             50  
3                    none          77             78             68  
4               completed          78             73             68

### method 2

In [38]:
df2.columns

Index(['gender', 'class group', 'parental level of education', 'lunch',
       'test preparation course', 'math score', 'reading score',
       'writing score'],
      dtype='object')

In [39]:
df2.columns = ['sex', 'class group', 'parental education', 'lunch',
       'test preparation course', 'math score', 'reading score',
       'writing score']
df2.columns

Index(['sex', 'class group', 'parental education', 'lunch',
       'test preparation course', 'math score', 'reading score',
       'writing score'],
      dtype='object')

### method 3 - using replace and editing the string

In [40]:
df.columns

Index(['gender', 'class group', 'parental level of education', 'lunch',
       'test preparation course', 'math score', 'reading score',
       'writing score'],
      dtype='object')

In [41]:
df.columns = df.columns.str.replace(' ', '_')
df.columns

Index(['gender', 'class_group', 'parental_level_of_education', 'lunch',
       'test_preparation_course', 'math_score', 'reading_score',
       'writing_score'],
      dtype='object')

In [ ]:
df.columns = df.columns.str.strip() # use it if you have white spaces before or after col names

### Direction of Calculations

In [42]:
df.head()

gender class_group parental_level_of_education         lunch  \
0    male     group A                 high school      standard   
1  female     group D            some high school  free/reduced   
2    male     group E                some college  free/reduced   
3    male     group B                 high school      standard   
4    male     group E          associate's degree      standard   

  test_preparation_course  math_score  reading_score  writing_score  
0               completed          67             67             63  
1                    none          40             59             55  
2                    none          59             60             50  
3                    none          77             78             68  
4               completed          78             73             68

In [43]:
df[['math_score', 'writing_score']].mean()

math_score       66.396
writing_score    67.738
dtype: float64

In [44]:
df.mean(numeric_only=True)

math_score       66.396
reading_score    69.002
writing_score    67.738
dtype: float64

In [46]:
# calculate the average of all 3 scores per student and place it a new column

df['average_score'] = round(df.mean(numeric_only=True, axis=1),2)
df.head()

gender class_group parental_level_of_education         lunch  \
0    male     group A                 high school      standard   
1  female     group D            some high school  free/reduced   
2    male     group E                some college  free/reduced   
3    male     group B                 high school      standard   
4    male     group E          associate's degree      standard   

  test_preparation_course  math_score  reading_score  writing_score  \
0               completed          67             67             63   
1                    none          40             59             55   
2                    none          59             60             50   
3                    none          77             78             68   
4               completed          78             73             68   

   average_score  
0          65.67  
1          51.33  
2          56.33  
3          74.33  
4          73.00

In [ ]:
df['average_score'] = round(df[['reading_score', 'writing_score']].mean(numeric_only=True, axis=1),2)
df.head()

### Aggregation

In [47]:
df['math_score'].mean()

66.396

In [48]:
# get the average of math score by gender 

dfAvgGender = df[['gender','math_score' ]].groupby('gender').mean()
dfAvgGender

math_score
gender            
female   63.196687
male     69.384913

In [50]:
# advanced aggregation
# use average for math score and min for reading score
df.groupby('parental_level_of_education').agg({'math_score':'mean'
                                               , 'reading_score':'min'})

math_score  reading_score
parental_level_of_education                           
associate's degree            69.536946             34
bachelor's degree             71.491071             34
high school                   65.207921             28
master's degree               71.585714             40
some college                  65.297297             27
some high school              60.701571             31

In [56]:
# use a custom function 
# for example diff between min and max

def MinMaxDiff(x):
    return x.max() - x.min()

#apply the function in agg

df.groupby('parental_level_of_education')[['math_score']].agg(MinMaxDiff)

math_score
parental_level_of_education            
associate's degree                   77
bachelor's degree                    65
high school                          71
master's degree                      62
some college                         87
some high school                     72